src/state/sagas/datasets/PollTwingraphStatus/PollTwingraphStatus.js

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Cosmo Tech.
+// Licensed under the MIT license.
+
+import { call, put, takeEvery, delay } from 'redux-saga/effects';
+import { t } from 'i18next';
+import { Api } from '../../../../services/config/Api';
+import { DATASET_ACTIONS_KEY } from '../../../commons/DatasetConstants';
+import { TWINGRAPH_STATUS } from '../../../../services/config/ApiConstants';
+import { TWINGRAPH_STATUS_POLLING_DELAY } from '../../../../services/config/FunctionalConstants';
+import { dispatchSetApplicationErrorMessage } from '../../../dispatchers/app/ApplicationDispatcher';
+
+export function* pollTwingraphStatus(action) {
+  let twingraphStatus = TWINGRAPH_STATUS.PENDING;
+  const datasetId = action.datasetId;
+  const organizationId = action.organizationId;
+
+  try {
+    while (![TWINGRAPH_STATUS.READY, TWINGRAPH_STATUS.ERROR, TWINGRAPH_STATUS.UNKNOWN].includes(twingraphStatus)) {
+      const { data: newStatus } = yield call(Api.Datasets.getDatasetTwingraphStatus, organizationId, datasetId);
+      if ([TWINGRAPH_STATUS.READY, TWINGRAPH_STATUS.ERROR, TWINGRAPH_STATUS.UNKNOWN].includes(newStatus)) {
+        twingraphStatus = newStatus;
+        yield put({
+          type: DATASET_ACTIONS_KEY.UPDATE_DATASET,
+          datasetId,
+          datasetData: { status: newStatus },
+        });
+      }
+      yield delay(TWINGRAPH_STATUS_POLLING_DELAY);
```

Comment:
I wonder if we could use a `do...while` here instead, but we can still change it later.
The `yield delay` can be put in an `else` block though

Reply:
Good idea, we will always have at least one iteration, I refactor this :wink: